src/Conversion/ONNXToKrnl/ControlFlow/Loop.cpp

```diff
@@ -239,47 +266,57 @@ struct ONNXLoopOpLowering : public ConversionPattern {
     // accumulate dynamic tensor
     // Convert the memref<?xmemref<>> to a memref
     SmallVector<Value, 4> newOutputs;
-    for (auto output : outputs) {
+    for (unsigned long i = 0; i < outputs.size(); i++) {
+      Value output = outputs[i];
       auto seqElementType =
           output.getType().cast<MemRefType>().getElementType();
       if (seqElementType.isa<MemRefType>()) {
-        // need to convert memref<memrefs<xT>> to memref<xT>
-        // TODO: need a IF statement to handle output is empty
-        // we can safely give 0 to the dynamic dim for alloc
-        // Here loop is assumed to be executed at least once.
-        auto firstElement =
-            create.krnl.load(output, create.math.constantIndex(0));
-        SmallVector<mlir::Value, 4> allocParams;
-        SmallVector<int64_t, 4> dims;
-        dims.emplace_back(output.getType().cast<MemRefType>().getShape()[0]);
-        if (output.getType().cast<MemRefType>().getShape()[0] == -1)
-          allocParams.emplace_back(create.mem.dim(output, 0));
-        for (auto i = 0;
-             i < firstElement.getType().cast<MemRefType>().getRank(); i++) {
-          dims.emplace_back(
-              firstElement.getType().cast<MemRefType>().getShape()[i]);
-          if (firstElement.getType().cast<MemRefType>().getShape()[i] == -1)
-            allocParams.emplace_back(create.mem.dim(firstElement, i));
+        // need to distinguish seqType in v_final and scan
+        if (i < loopOp.v_final().size()) {
+          // In v_final
+          auto v = create.krnl.load(output, zero);
+          newOutputs.emplace_back(v);
+        } else {
+          // scan output
+          // need to convert memref<memrefs<xT>> to memref<xT>
+          // TODO: need a IF statement to handle output is empty
+          // we can safely give 0 to the dynamic dim for alloc
+          // Here loop is assumed to be executed at least once.
+          auto firstElement =
+              create.krnl.load(output, create.math.constantIndex(0));
+          SmallVector<mlir::Value, 4> allocParams;
+          SmallVector<int64_t, 4> dims;
+          dims.emplace_back(output.getType().cast<MemRefType>().getShape()[0]);
+          if (output.getType().cast<MemRefType>().getShape()[0] == -1)
+            allocParams.emplace_back(create.mem.dim(output, 0));
+          for (auto i = 0;
+               i < firstElement.getType().cast<MemRefType>().getRank(); i++) {
```

Comment:
FYI, we recently add some utility functions to get shape and element type in [TypeUtilities.hpp](https://github.com/onnx/onnx-mlir/blob/main/src/Support/TypeUtilities.hpp). So you can call `getShape(firstElement.getType())` which will check and cast the type for us.

Reply:
Useful utility, but not fit this situation: construct shape from two shape along with values for allocation.